### PR TITLE
[feature/batch_send] Add send_notification_batch

### PR DESCRIPTION
**Pull request by matangover**

### CODE REVIEW

File: apns2/client.py [hunk 1]
```diff
@@ -1,39 +1,149 @@
...
+            ensure_ascii=False,
+            separators=(',', ':')
+        ).encode('utf-8')
 
-        headers = {
-            'apns-priority': priority.value
-        }
-        if topic:
-            headers['apns-topic'] = topic
+        headers = {'apns-topic': topic}
```
Comment (mccomplete) [inline apns2/client.py:58]: Why this change?

File: apns2/client.py [hunk 2]
```diff
@@ -1,39 +1,149 @@
...
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        '''
+        Send a notification to a list of tokens in batch. Instead of sending a synchronous request
+        for each token, send multiple requests concurrently. This is done on the same connection,
+        using HTTP/2 streams (one request per stream).
+
+        APNs allows many streams simultaneously, but the number of streams can vary depending on
+        server load. This method reads the SETTINGS frame sent by the server to figure out the
```
Comment (mccomplete) [inline apns2/client.py:88]: What's the difference between this and SAFETY_MAXIMUM?

File: apns2/client.py [hunk 3]
```diff
@@ -1,39 +1,149 @@
...
+        # When reaching the maximum concurrent streams limit, wait for a response before sending
+        # another request.
+        while open_streams or next_token:
+            # Update the max_concurrent_streams on every iteration since a SETTINGS frame can be
+            # sent by the server at any time.
+            self.update_max_concurrent_streams()
+            if next_token and len(open_streams) < self._max_concurrent_streams:
+                logger.info('Sending to token %s', next_token)
+                stream_id = self.send_notification_async(next_token, notification, topic, priority)
+                open_streams.append((stream_id, next_token))
```
Comment (mccomplete) [inline apns2/client.py]: I want a named tuple.  But you knew I was going to say that.

File: apns2/client.py [hunk 4]
```diff
@@ -1,39 +1,149 @@
...
+
+                try:
+                    next_token = next(token_iterator)
+                except StopIteration:
```
Comment (mccomplete) [inline apns2/client.py:122]: It's _so_ weird that this is an exception.

File: apns2/client.py [hunk 5]
```diff
@@ -1,39 +1,149 @@
...
+        # frame before starting to send notifications.
+        self._connection.connect()
+        result_counter = collections.Counter()
+        open_streams = collections.deque()
+        token_iterator = iter(tokens)
+        next_token = next(token_iterator)
+        # Loop on the tokens, sending as many requests as possible concurrently to APNs.
+        # When reaching the maximum concurrent streams limit, wait for a response before sending
+        # another request.
+        while open_streams or next_token:
```
Comment (mccomplete) [inline apns2/client.py]: I usually prefer not to rely on empty containers being falsy.

File: apns2/client.py [hunk 6]
```diff
@@ -1,39 +1,149 @@
...
+        APNs allows many streams simultaneously, but the number of streams can vary depending on
+        server load. This method reads the SETTINGS frame sent by the server to figure out the
+        maximum number of concurrent streams. Typically, APNs reports a maximum of 500.
+        '''
+        # Make sure we're connected to APNs, so that we receive and process the server's SETTINGS
+        # frame before starting to send notifications.
+        self._connection.connect()
+        result_counter = collections.Counter()
+        open_streams = collections.deque()
+        token_iterator = iter(tokens)
```
Comment (mccomplete) [inline apns2/client.py]: I can't help wondering if you could refactor this to use a for loop over tokens, instead of an iterator object.  I find Python's iterator API kind of ugly, although nowhere near as ugly as the old Java/C++ iterator APIs good riddance to them.

File: apns2/client.py [hunk 7]
```diff
@@ -1,39 +1,149 @@
...
+        next_token = next(token_iterator)
```
Comment (mccomplete) [inline apns2/client.py]: How do you know this won't raise StopIteration?

File: apns2/client.py [hunk 8]
```diff
@@ -1,39 +1,149 @@
...
+                try:
+                    next_token = next(token_iterator)
+                except StopIteration:
+                    # No tokens remaining. Proceed to get results for pending requests.
+                    logger.info('Finished sending all tokens, waiting for pending requests.')
+                    next_token = None
+            else:
+                # We have at least one request waiting for response (otherwise we would have either
+                # sent new requests or exited the while loop.)
+                assert open_streams
```
Comment (mccomplete) [inline apns2/client.py]:
``` python
assert len(open_streams) > 0
```

File: test/test_client.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+from apns2.payload import Payload
+
+class ClientTestCase(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        logging.basicConfig()
+        cls.tokens = ["%064x" % i for i in range(10000)]
+        cls.notification = Payload(alert="Hello, world!")
+        cls.topic = "com.example.App"
+
```
Comment (mccomplete) [inline test/test_client.py:18]: Too many newlines

File: test/test_client.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    
+    def setUp(self):
+        self.open_streams = 0
+        self.max_open_streams = 0
+        self.mock_results = None
+        
+        with patch("apns2.client.HTTP20Connection") as mock_connection_constructor, \
+            patch("apns2.client.init_context"):
+            self.mock_connection = Mock()
+            self.mock_connection.get_response.side_effect = self.mock_get_response
```
Comment (mccomplete) [inline test/test_client.py:28]: Could you do this with a good old class instead of the mock library?

File: test/test_client.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            self.mock_connection.request.side_effect = self.mock_request
+            self.mock_connection.remote_settings.max_concurrent_streams = 500
```
Comment (mccomplete) [inline test/test_client.py:30]: Shouldn't you be testing different possible values of max_concurrent_streams?

File: test/test_client.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            
+    def mock_get_response(self, *dummy_args):
+        self.open_streams -= 1
+        if self.mock_results:
+            reason = self.mock_results[self.mock_connection.get_response.call_count - 1]
+            response = Mock(status=200 if reason == "Success" else 400)
+            response.read.return_value = '{"reason": "%s"}' % reason
+            return response
+        else:
+            return Mock(status=200)
```
Comment (mccomplete) [inline test/test_client.py:42]: What does it mean if get_response() is called when mock_results is falsy?

File: test/test_client.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+            self.mock_connection.get_response.side_effect = self.mock_get_response
+            self.mock_connection.request.side_effect = self.mock_request
+            self.mock_connection.remote_settings.max_concurrent_streams = 500
+            mock_connection_constructor.return_value = self.mock_connection
+            self.client = APNsClient(cert_file=None)
+            
+    def mock_get_response(self, *dummy_args):
+        self.open_streams -= 1
+        if self.mock_results:
+            reason = self.mock_results[self.mock_connection.get_response.call_count - 1]
```
Comment (mccomplete) [inline test/test_client.py]: Maybe it would be better to maintain a dict of tokens to responses, to make sure you get the right response for the right token?

**Comment by mccomplete** [comment]
Round 1 done.

File: apns2/client.py (hunk 1 above)
Comment (matangover) [inline apns2/client.py:58]: Topic is required. Sending to APNs without a topic returns an error.

Comment (matangover) [inline apns2/client.py:58]: Also, there's no reason to send priority if it's set to the default value. It just wastes time and bandwidth.

File: apns2/client.py (hunk 2 above)
Comment (matangover) [inline apns2/client.py:88]: I don't understand the question.

File: apns2/client.py (hunk 6 above)
Comment (matangover) [inline apns2/client.py]: I don't think that's possible, because the loop doesn't only iterate over all the tokens, it also iterates over all the requests.

File: apns2/client.py (hunk 4 above)
Comment (matangover) [inline apns2/client.py:122]: Not my fault.

File: apns2/client.py (hunk 7 above)
Comment (matangover) [inline apns2/client.py]: Thanks, fixed and added test.

File: apns2/client.py (hunk 8 above)
Comment (matangover) [inline apns2/client.py]: Done

File: apns2/client.py (hunk 5 above)
Comment (matangover) [inline apns2/client.py]: ok, made it explicit.

File: apns2/client.py (hunk 3 above)
Comment (matangover) [inline apns2/client.py]: Done.

File: test/test_client.py (hunk 13 above)
Comment (matangover) [inline test/test_client.py]: The tests doesn't rely on the order of responses so that is an unneeded complication imo.

File: test/test_client.py (hunk 9 above)
Comment (matangover) [inline test/test_client.py:18]: Fixed

File: test/test_client.py (hunk 10 above)
Comment (matangover) [inline test/test_client.py:28]: Why would I want to do that? My reasons for using the Mock library are:
1. Inline, one-line generation of mocks. Note that I have more than one mocked class here. I would need at least three classes: MockConnection, MockSettings, MockResponse.
2. Mock gives me useful things like call_count, assert_called_with, side_effect.
3. mock.patch is very useful, with automatic undo of the patch when finished.

File: test/test_client.py (hunk 11 above)
Comment (matangover) [inline test/test_client.py:30]: This is a default that's overridden in `test_send_notification_batch_overrides_server_max_concurrent_streams_if_too_small` and `test_send_notification_batch_overrides_server_max_concurrent_streams_if_too_large`.

File: test/test_client.py (hunk 12 above)
Comment (matangover) [inline test/test_client.py:42]: `mock_results` is way for a test to say "get_response mock, please return the following values [different value every time you're called". In all other cases, when `mock_results` is `None` (which is the case in all tests but `test_send_notification_batch_reports_different_results`), get_response always returns success.

**Comment by matangover** [comment]
Round 2 done.

File: test/test_client.py (hunk 13 above)
Comment (mccomplete) [inline test/test_client.py]: But wouldn't it be nice to test if this works correctly?  Usually we don't test log messages, but in this case, we would like the log messages to report the correct outcome for each token.

File: test/test_client.py (hunk 10 above)
Comment (mccomplete) [inline test/test_client.py:28]: Sorry, I guess what I meant was to use a good old class for mock_connection.  With mock_connection, you're defining methods and using instance variables; I think it might be much more straightforward with a good old class.  The Mock library probably makes sense for the other mocks.

Also, I hate the mock.side_effect attribute.  I think I'm almost used to reading it and using it by now, but I still think it's confusing and counterintuitive.

File: apns2/client.py [hunk 14]
```diff
@@ -1,39 +1,156 @@
...
+        self,
+        tokens,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        '''
+        Send a notification to a list of tokens in batch. Instead of sending a synchronous request
+        for each token, send multiple requests concurrently. This is done on the same connection,
+        using HTTP/2 streams (one request per stream).
```
Comment (mccomplete) [inline apns2/client.py:85]: I'm not sure this is the best interface, specifically in terms of the return value, which only returns counts instead of returning a dict of token: outcome.  The caller might want to know which tokens were unregistered, for example.

You could also change this method so it returns a dict, and define a wrapper method that calls this method, logs the results, and returns the outcome counts.

**Comment by mccomplete** [comment]
Round 3 done.

File: apns2/client.py (hunk 14 above)
Comment (matangover) [inline apns2/client.py:85]: Good idea. I changed it to return a dict. The helper method isn't needed right now.

File: test/test_client.py (hunk 13 above)
Comment (matangover) [inline test/test_client.py]: You're right. I changed the tests to test the per-token results and also made the tests track the stream_id to make sure a token's [mock] response corresponds to its [mock] request.

**Comment by mccomplete** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;SCtkGu5q&#x2F;94-modify-notificatricks-code-to-use-new-send-notification-batch-in-our-fork-of-pyapns2"> Modify notificatricks code to use new send_notification_batch in our fork of PyAPNs2</a>

**Comment by matangover** [comment]
Round 4 done. (2 months ago.)
The reason I didn't post "round done" till now is that I wanted to implement better error handling in send_notification_batch, but it appears to be impossible. Let's discuss this f2f.

File: test/test_client.py [hunk 15]
```diff
@@ -43,10 +45,14 @@ def mock_request(self, *dummy_args):
         self.open_streams += 1
         if self.open_streams > self.max_open_streams:
             self.max_open_streams = self.open_streams
+        stream_id = self.next_stream_id
```
Comment (mccomplete) [inline test/test_client.py:49]: newline after if

File: test/test_client.py [hunk 16]
```diff
@@ -68,13 +74,25 @@ def test_send_notification_batch_reports_different_results(self):
...
+    def test_connect_retries_failed_connection(self):
+        self.mock_connection.connect.side_effect = [RuntimeError, RuntimeError, None]
+        self.client.connect()
+        self.assertEqual(self.mock_connection.connect.call_count, 3)
+    
+    def test_connect_stops_on_reaching_max_retries(self):
+        self.mock_connection.connect.side_effect = [RuntimeError] * 4
+        with self.assertRaises(ConnectionError):
+            self.client.connect()
+        self.assertEqual(self.mock_connection.connect.call_count, 3)
```
Comment (mccomplete) [inline test/test_client.py:99]: newline after with

**Comment by mccomplete** [comment]
Round 5 done.  A few small comments.

**Comment by matangover** [comment]
Round 6 done.
